gnat2goto/driver/tree_walk.adb

```diff
@@ -1007,113 +984,311 @@ package body Tree_Walk is
+         Append_Op (Result_Block,
+                    Make_Code_Assign (Rhs             =>
+                                  Typecast_If_Necessary (Zero, CProver_Size_T),
+                                      Lhs             => Sum_Size_Var,
+                                      Source_Location => Source_Loc));
+         for I in Slices'Range loop
+            Build_Sum_Size (Slices (I));
+         end loop;
+         Append_Op (Result_Block,
+                    Make_Code_Assign (Rhs             => Dest_Temp_Alloc,
+                                      Lhs             => Dest_Temp,
+                                      Source_Location => Source_Loc));
+         for I in Slices'Range loop
+            Process_Slice (Slices (I));
+         end loop;
+         Append_Op (Result_Block,
+                    Make_Code_Assign (Rhs             => Memcpy_Dest,
+                                      Lhs             => Memcpy_Lhs,
+                                      Source_Location => Source_Loc));
+         return Result_Block;
+      end Build_Concat_Assign_Body;
+
+      Func_Symbol : constant Symbol :=
+        Build_Function (Name           => Function_Name,
+                        RType          => Ret_Type,
+                        Func_Params    => Concat_Params,
+                        FBody          => Build_Concat_Assign_Body,
+                        A_Symbol_Table => Global_Symbol_Table);
+
+      Func_Call : constant Irep :=
+        Make_Side_Effect_Expr_Function_Call (
+                                  Arguments       => Concat_Arguments,
+                                  I_Function      => Symbol_Expr (Func_Symbol),
+                                  Source_Location => Source_Loc,
+                                  I_Type          => Ret_Type);
+      Concat_Lhs : constant Irep :=
+        Fresh_Var_Symbol_Expr (Ret_Type, "concat_Lhs");
+   begin
+      Append_Argument (Concat_Arguments,
+                       Do_Expression (LHS_Node));
+      for I in RHS_Arrays'Range loop
+         Append_Argument (Concat_Arguments,
+                          RHS_Arrays (I));
+      end loop;
 
-      Append_Op (Ret,
-                 Make_Code_Function_Call (Arguments       => Copy_Args,
-                                          I_Function      => Copy_Func,
-                                          Lhs             => LHS_fun_call,
-                                          Source_Location => Sloc (N),
-                                          I_Type          => Make_Void_Type));
+      return Make_Code_Assign (Rhs             => Func_Call,
+                               Lhs             => Concat_Lhs,
+                               Source_Location => Source_Loc);
+   end Do_Array_Assignment;
 
-      return Ret;
+   -------------------------
+   -- Do_RHS_Array_Assign --
+   -------------------------
 
-   end Do_Array_Assignment;
+   function Do_RHS_Array_Assign (N : Node_Id) return Irep_Array
+   is
+   begin
```

Comment:
I'm not sure I like the design of this. Basically, this creates a special case just for op-concat, and we can't use this in `Do_Expression` because it invokes `Do_Expression` itself?

Reply:
Well we could (and will in the next PR) use it in the `Do_Expression` but now it returns an array of ireps and `Do_Expression` expects just one.